### PR TITLE
STM32: Add base support for stm32u5 series and nucleo_u575zi_q

### PR DESCRIPTION
**Pull request by erwango**
Add basic support for stm32u5 based socs:
- soc: base + STM32U575 support
- dts: base *.dtsi + STM32U5 clock_control bindings
- drivers: Includes updates on the following drivers: clock_control, gpio, interrupt_controller, uart

Additionally, provide basic support for nucleo_u575zi_q

### CODE REVIEW

**Comment by erwango** [comment]
I'm getting 2 issues on compliance checks:
First one is the following:

```
-:581: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#581: FILE: drivers/clock_control/clock_stm32_ll_u5.c:48:
+void config_pll_init(LL_UTILS_PLLInitTypeDef *pllinit)
                                              ^
```
This is a known false positive, documented here: https://docs.zephyrproject.org/2.5.0/development_process/dev_env_and_tools.html#continuous-integration
Documentation suggests the following: " Maintainers can override the -1 in cases where the CI infrastructure gets the wrong answer".

Second one is based of usage of `IS_ENABLED` using non Kconfig symbol:
```
-:907: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
#907: FILE: drivers/clock_control/clock_stm32_ll_u5.c:374:
+	if (IS_ENABLED(STM32_HSE_BYPASS)) {

-:954: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
#954: FILE: drivers/clock_control/clock_stm32_ll_u5.c:421:
+		if (IS_ENABLED(STM32_HSE_BYPASS)) {
```
This is actually already in use in the code, mostly in `include/devicetree.h`:
https://github.com/zephyrproject-rtos/zephyr/blob/2c606239e2e0768c9345ad88cd7ac13621f4feba/include/devicetree.h#L1322-L1323


@galak is that fine to consider these as false positive ?

File: drivers/clock_control/clock_stm32_ll_u5.c
Comment (ABOSTM) [inline drivers/clock_control/clock_stm32_ll_u5.c:35]: HCLK is the core clock it is also SystemCoreClock. So this restriction doesn't apply in this case.

File: boards/arm/nucleo_u575zi_q/doc/nucleou575zi_q.rst
Comment (ABOSTM) [inline boards/arm/nucleo_u575zi_q/doc/nucleou575zi_q.rst:181]: According to dtsi below:
- LD2 : PB7

File: boards/arm/nucleo_u575zi_q/nucleo_u575zi_q.yaml
Comment (ABOSTM) [inline boards/arm/nucleo_u575zi_q/nucleo_u575zi_q.yaml:10]: Only 192 have been declared as sram

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: dts/arm/st/u5/stm32u5.dtsi
Comment (erwango) [reply dts/arm/st/u5/stm32u5.dtsi:84]: Done

File: dts/bindings/clock/st,stm32u5-pll-clock.yaml
Comment (erwango) [reply dts/bindings/clock/st,stm32u5-pll-clock.yaml:60]: Right, and actually none odd value is allowed for PLL1.
I'm actually removing PLLP then as its usage is not implemented in clock_control driver

File: dts/arm/st/u5/stm32u5.dtsi
Comment (erwango) [reply dts/arm/st/u5/stm32u5.dtsi:102]: done

File: drivers/clock_control/clock_stm32_ll_u5.c
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_u5.c:122]: I'm not seeing any reference to ABH4 in Ref man or CMSIS file.
Can you provide more info ?

Comment (ABOSTM) [reply drivers/clock_control/clock_stm32_ll_u5.c:122]: image

File: soc/arm/st_stm32/stm32u5/soc.c
Comment (erwango) [reply soc/arm/st_stm32/stm32u5/soc.c:34]: Voltage scaling is set (line 62).
For EPOD, I'd expect to get it set as part of clock control.
Thought so far, it is not yet handled.

File: drivers/clock_control/clock_stm32_ll_u5.c
Comment (ABOSTM) [reply drivers/clock_control/clock_stm32_ll_u5.c:122]: My bad, it was in rev 0.1 but removed in rev1.0

File: drivers/clock_control/clock_stm32_ll_u5.c
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_u5.c:35]: Cf https://github.com/zephyrproject-rtos/zephyr/commit/77db273f6f84b3d94d34b5e72070ec6bc31e41cf
I don't see the difference in U5 vs other series

File: dts/arm/st/u5/stm32u575Xi.dtsi
Comment (erwango) [reply dts/arm/st/u5/stm32u575Xi.dtsi:11]: Added the full contiguous SRAM size

File: dts/arm/st/u5/stm32u5.dtsi
Comment (erwango) [reply dts/arm/st/u5/stm32u5.dtsi:72]: This is indeed awkward but correct.
'pll1' and first 'pll' are node names. Last 'pll' is node address.

**Comment by erwango** [comment]
@ABOSTM Thanks for the review, I think I addressed all of your comments.

File: drivers/clock_control/clock_stm32_ll_u5.c
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_u5.c:35]: As discussed, restriction doesn't apply and should be fixed on at least some other series as well;
Will be fixed globally once this PR is merged.

**Review by ABOSTM — CHANGES_REQUESTED** [review]
As discussed, need EPOD enabling above 55MHz, otherwise, LGTM

**Comment by erwango** [comment]
> As discussed, need EPOD enabling above 55MHz, otherwise, LGTM

@ABOSTM Added code to handle this. Please review once again

File: drivers/clock_control/clock_stm32_ll_u5.c
Comment (ABOSTM) [inline drivers/clock_control/clock_stm32_ll_u5.c:345]:
```suggestion
		 * Make sure prescaler value is not higher than 16
```
This is prescaler value it has no unit

File: drivers/clock_control/clock_stm32_ll_u5.c
Comment (ABOSTM) [inline drivers/clock_control/clock_stm32_ll_u5.c:358]: To be moved in `if` above: EPOD Booster should be enabled only if freq >=55MHz

**Review by ABOSTM — CHANGES_REQUESTED** [review]

**Review by ABOSTM — APPROVED** [review]
LGTM

File: soc/arm/st_stm32/stm32u5/soc.c
Comment (FRASTM) [inline soc/arm/st_stm32/stm32u5/soc.c:64]: Could this sequence be a little bit more commented (and the counter value of '100')
Is there any input from the arc/arm/include/aarch32/cortex_m/dwt.h ?

File: dts/arm/st/u5/stm32u575.dtsi
Comment (FRASTM) [inline dts/arm/st/u5/stm32u575.dtsi:2]: `Copyright (c) 2021` ?

File: dts/arm/st/u5/stm32u575Xi.dtsi
Comment (FRASTM) [inline dts/arm/st/u5/stm32u575Xi.dtsi:2]: `Copyright (c) 2021` ?

File: boards/arm/nucleo_u575zi_q/doc/nucleou575zi_q.rst
Comment (erwango) [reply boards/arm/nucleo_u575zi_q/doc/nucleou575zi_q.rst:241]: ok, strange, seems it has been removed for some reason.
I'll clear the link. Thanks

File: soc/arm/st_stm32/stm32u5/soc.c
Comment (erwango) [reply soc/arm/st_stm32/stm32u5/soc.c:64]: This is indeed covered by `timing_init` function, thanks for the catch.
This function is optional in zephyr (controlled by `TIMING_FUNCTIONS_NEED_AT_BOOT`), but looking into more details this also seems to be the case in Cube samples (only available in LL examples).
Clearing this part.

File: soc/arm/st_stm32/stm32u5/soc.c
Comment (erwango) [reply soc/arm/st_stm32/stm32u5/soc.c:16]: > If the stm32_ll_icache.h is a new file of ../modules/hal/stm32/stm32cube/common_ll/include
could you link the commit

It's part of cube update for u5:
https://github.com/zephyrproject-rtos/hal_stm32/pull/109/commits/1ef480ba5f556084a072d490969051aa07b1fddf#diff-6a48f59beb44075e26d8e138800b82ffe6cba4004bcad8536d1c159865a05b5b

> else it could use the the SCB_EnableICache() functions

This looks different with what is done in LL_ICACHE_Enable.
I'll dig into that. If similar, I agree that CMSIS functions should be preferred.

Comment (erwango) [reply soc/arm/st_stm32/stm32u5/soc.c:16]: After investigation, it seems that stm23u5 icache is a dedicated hardware block with specific registers ([Cf Ref man, chapter 8](https://www.st.com/resource/en/reference_manual/rm0456-stm32u575585-armbased-32bit-mcus-stmicroelectronics.pdf)).
So, I'm keeping the proposed implementation

**Comment by erwango** [comment]
@FRASTM Thanks for the review. Comments addressed

**Comment by galak** [comment]
>     * 
> 
> 
> @galak is that fine to consider these as false positive ?

Yes, I think its fine that these are false positives.

**Review by FRASTM — APPROVED** [review]

**Comment by galak** [comment]
@erwango just ping me when this is ready to merge and I'll merge it.

**Comment by cfriedt** [comment]
Looks like we have all of the approvals and we're just waiting on zephyrproject-rtos/hal_stm32#109 and zephyrproject-rtos/hal_stm32#110

**Comment by erwango** [comment]
> Looks like we have all of the approvals and we're just waiting on [zephyrproject-rtos/hal_stm32#109](https://github.com/zephyrproject-rtos/hal_stm32/pull/109) and [zephyrproject-rtos/hal_stm32#110](https://github.com/zephyrproject-rtos/hal_stm32/pull/110)

Just merged [zephyrproject-rtos/hal_stm32#109](https://github.com/zephyrproject-rtos/hal_stm32/pull/109) and updated west.yml.
[zephyrproject-rtos/hal_stm32#110](https://github.com/zephyrproject-rtos/hal_stm32/pull/110) is part of another task (update hal_stm32 module with latest STM32Cube packages updates, which we do once per release).

**Comment by cfriedt** [comment]
Whitespace 😭

**Comment by erwango** [comment]
> @erwango - just some whitespace errors :-/

@cfriedt, are you referring to error reported by checkpatch?
```
-:577: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#577: FILE: drivers/clock_control/clock_stm32_ll_u5.c:50:
+static void config_pll_init(LL_UTILS_PLLInitTypeDef *pllinit)
                                                     ^
```

If yes, as mentioned [here](https://github.com/zephyrproject-rtos/zephyr/pull/37139#issuecomment-885040930), these are false documented false positive, which was [acknowlegded by @galak](https://github.com/zephyrproject-rtos/zephyr/pull/37139#issuecomment-887578054).
If you're referring to something else, would you mind pointing it to me :-) ?

**Comment by galak** [comment]
I've merged this.

**Comment by jli157** [comment]
@erwango when will TF-M be enabled for STM32U5 in Zephyr project?

**Comment by erwango** [comment]
> @erwango when will TF-M be enabled for STM32U5 in Zephyr project?

@jli157, I didn't had time to check this yet, so far U5 support is not complete in TF-M upstream yet.
This should be completed for TF-M v1.4 release (Mid-october).
Then, we'll need to wait TF-M v1.4 is integrated with Zephyr.

**Comment by jli157** [comment]
> > @erwango when will TF-M be enabled for STM32U5 in Zephyr project?
> 
> @jli157, I didn't had time to check this yet, so far U5 support is not complete in TF-M upstream yet. This should be completed for TF-M v1.4 release (Mid-october). Then, we'll need to wait TF-M v1.4 is integrated with Zephyr.

@erwango thanks for the update!

**Pull request merged**
